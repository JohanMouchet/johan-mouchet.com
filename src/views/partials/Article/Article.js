// @flow

import * as React from "react";
import cx from "classnames";
import "./Article.scss";

type Props = {
  size?: "sm" | "lg",
  margin?: "no-margin" | "no-top-margin" | "no-bottom-margin",
  variant?: "secondary",
  className?: string | Array<string> | Object,
  children: Node,
};

const Article = ({ size, margin, variant, className, children }: Props) => {
  const classNames = cx(
    "p-article",
    size && `p-article--${size}`,
    margin && `p-article--${margin}`,
    variant && `p-article--${variant}`,
    variant && size && `p-article--${variant}-${size}`,
    className
  );

  return <article className={classNames}>{children}</article>;
};

export default Article;
