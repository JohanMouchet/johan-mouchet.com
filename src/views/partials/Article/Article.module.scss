@import "assets/styles/settings/_settings.module";
@import "assets/styles/tools/_tools.module";

.p-article {
  $p-article: (
    v-padding: t-vertical-rhythm(2),
    scale-factor: 2,
  );

  position: relative;
  margin-top: map-get($p-article, v-padding);
  margin-bottom: map-get($p-article, v-padding);

  /** Margin **/
  &--sm {
    margin-top: map-get($p-article, v-padding) /
      map-get($p-article, scale-factor);
    margin-bottom: map-get($p-article, v-padding) /
      map-get($p-article, scale-factor);
  }

  &--lg {
    margin-top: map-get($p-article, v-padding) *
      map-get($p-article, scale-factor);
    margin-bottom: map-get($p-article, v-padding) *
      map-get($p-article, scale-factor);
  }

  &--no-margin {
    margin-top: 0;
    margin-bottom: 0;
  }

  &--no-top-margin {
    margin-top: 0;
  }

  &--no-bottom-margin {
    margin-bottom: 0;
  }

  /** Secondary **/
  &--secondary {
    padding-top: map-get($p-article, v-padding);
    padding-bottom: map-get($p-article, v-padding);
    background: $s-color-gray-lighter;

    @if lightness($s-color-gray-lighter) < 75 {
      h1,
      h2,
      h3,
      h4,
      h5,
      h6 {
        color: $s-color-text-dimmed;
      }

      p,
      span {
        color: $s-color-white;
      }
    }
  }

  &--secondary-sm {
    padding-top: map-get($p-article, v-padding) /
      map-get($p-article, scale-factor);
    padding-bottom: map-get($p-article, v-padding) /
      map-get($p-article, scale-factor);
  }

  &--secondary-lg {
    padding-top: map-get($p-article, v-padding) *
      map-get($p-article, scale-factor);
    padding-bottom: map-get($p-article, v-padding) *
      map-get($p-article, scale-factor);
  }
}
