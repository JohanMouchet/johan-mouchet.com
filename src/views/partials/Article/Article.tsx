import clsx, { ClassValue } from "clsx";
import styles from "./Article.module.scss";

export const Article = ({
  size,
  margin,
  variant,
  children,
  className,
  ...props
}: {
  size?: "sm" | "lg";
  margin?: "no-margin" | "no-top-margin" | "no-bottom-margin";
  variant?: "secondary";
  children: React.ReactNode;
  className?: ClassValue;
} & Omit<React.HTMLProps<HTMLDivElement>, "size">) => {
  const classNames = clsx(
    styles["p-article"],
    size && styles[`p-article--${size}`],
    margin && styles[`p-article--${margin}`],
    variant && styles[`p-article--${variant}`],
    variant && size && styles[`p-article--${variant}-${size}`],
    className
  );

  return (
    <article className={classNames} {...props}>
      {children}
    </article>
  );
};
