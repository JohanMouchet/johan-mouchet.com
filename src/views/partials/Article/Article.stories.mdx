import { Canvas, Meta, Story } from "@storybook/addon-docs/blocks";
import { select, text } from "@storybook/addon-knobs";
import Article from ".";

<Meta title="Partials/Article" component={Article} />

# Article

<Canvas>
  <Story name="default">
    <div style={{ outline: "1px dotted" }}>
      <Article
        size={select("size", { "": null, sm: "sm", lg: "lg" })}
        margin={select("margin", {
          "": null,
          "no-margin": "no-margin",
          "no-top-margin": "no-top-margin",
          "no-bottom-margin": "no-bottom-margin",
        })}
        variant={select("variant", { "": null, secondary: "secondary" })}
      >
        {text("children", "children")}
      </Article>
    </div>
  </Story>
</Canvas>
