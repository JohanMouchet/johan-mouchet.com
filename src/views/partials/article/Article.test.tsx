import { render } from "@testing-library/react";
import { expect, it } from "vitest";
import { Article } from "./Article";

it("Renders <Article /> without crashing", () => {
  expect(
    render(
      <Article size="sm" margin="no-margin" variant="secondary">
        children
      </Article>
    ).container.firstChild
  ).toMatchSnapshot();
});
