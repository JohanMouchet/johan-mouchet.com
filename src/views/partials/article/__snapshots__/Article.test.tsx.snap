// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Renders <Article /> without crashing 1`] = `
<article
  class="_p-article_63fc43 _p-article--sm_63fc43 _p-article--no-margin_63fc43 _p-article--secondary_63fc43 _p-article--secondary-sm_63fc43"
>
  children
</article>
`;
